java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/DictionaryBuilder.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.dictionary;
+
+import org.apache.arrow.algorithm.sort.VectorValueComparator;
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * The class for building a dictionary.
+ * It is intended for the scenario which can be frequently encountered in practice:
+ * the dictionary is not known a priori, so it is generated dynamically.
+ * In particular, when a new value arrives, it is tested to check if it is already
+ * in the dictionary. If so, it is simply neglected, otherwise, it is added to the dictionary.
+ *
+ * <p>
+ *   When all values have been evaluated, the dictionary can be considered complete.
+ *   So encoding can start afterward.
+ * </p>
+ * @param <V> the dictionary vector type.
+ */
+public abstract class DictionaryBuilder<V extends ValueVector> {
```

Comment:
I'd prefer not to create class hierarchies for only one implementation.  If there are plans to add additional builders, lets do that in follow-up PRs where we can assess the commonalities.  for now can you merge this into the TreeSet Builder below?

Reply:
Sure. Maybe we can extract a base class later, if it is necessary.